src/utilities/ParseExtra.cpp

```diff
@@ -69,36 +78,86 @@ namespace Utilities
+                    std::vector<uint8_t> nonceData;
+
+                    std::copy(it + 1 + readNonceSize, it + 1 + readNonceSize + nonceSize, std::back_inserter(nonceData));
+
+                    /* Loop through the nonce data looking for fields */
+                    for (auto is = nonceData.begin(); is != nonceData.end(); is++)
+                    {
+                        const uint8_t s = *is;
+
+                        const auto nElementsRemaining = std::distance(is, nonceData.end());
 
-                Crypto::Hash paymentIDHash;
+                        /* If we encounter a Payment ID field and there are enough bytes remaining in
+                           the nonce data and we have not encountered a payment ID, then read it out */
+                        if (s == Constants::TX_EXTRA_PAYMENT_ID_IDENTIFIER && nElementsRemaining > 32 && !seenPaymentID)
+                        {
+                            /* Grab the payment ID hash out of the nonce data */
+                            Crypto::Hash paymentIDHash;
 
-                /* Copy the payment ID into the hash */
-                std::copy(dataBegin, dataBegin + 32, std::begin(paymentIDHash.data));
+                            std::copy(is + 1, is + 1 + 32, std::begin(paymentIDHash.data));
 
-                /* Convert to a string */
-                std::string paymentID = Common::podToHex(paymentIDHash);
+                            /* Assign the payment ID to the parsed result */
+                            parsed.paymentID = Common::podToHex(paymentIDHash);
 
-                /* Convert it to lower case */
-                std::transform(paymentID.begin(), paymentID.end(), paymentID.begin(), ::tolower);
+                            seenPaymentID = true;
 
-                parsed.paymentID = paymentID;
+                            /* Advance the iterator by the tag (1-byte) + the hash size (32-bytes) */
+                            advanceIterator += 1 + 32;
 
-                /* Advance past the payment ID */
-                it += 32 + 1 + 1;
+                            continue;
+                        }
 
-                seenPaymentID = true;
+                        /* If we encounter an arbitrary data field and there is at least one byte remaining
+                           and we have not encountered the arbitrary data yet, then read it out */
+                        if (s == Constants::TX_EXTRA_ARBITRARY_DATA_IDENTIFIER && nElementsRemaining > 1 && !seenExtraData)
+                        {
+                            /* Read out the size of the data */
+                            size_t dataSize = 0;
+
+                            const auto readDataSize = Tools::read_varint(is + 1, nonceData.end(), dataSize);
+
+                            /* If there are enough bytes left to read based upon the size above then
+                               read out the data */
+                            if (nElementsRemaining >= 1 + dataSize)
```

Comment:
Shouldn't this be 
```suggestion
                            if (nElementsRemaining >= 1 + dataSize + readDataSize)
```
to include the length of the varint?

Reply:
Yes, fixed locally, will push as others are addressed.

Reply:
Pushed